tfhe/src/integer/bigint/algorithms.rs

```diff
@@ -182,25 +223,89 @@ where
     (quotient, remainder)
 }
 
+pub(crate) fn slow_div_signed<T>(numerator: T, divisor: T) -> (T, T)
+where
+    T: Numeric
+        + ShlAssign<u32>
+        + Shl<u32, Output = T>
+        + Shr<u32, Output = T>
+        + BitOrAssign<T>
+        + SubAssign<T>
+        + BitAnd<T, Output = T>
+        + Ord
+        + Neg<Output = T>
+        + Add<T, Output = T>
+        + Sub<T, Output = T>,
+{
+    assert!(divisor != T::ZERO);
+
+    let positive_numerator = if numerator < T::ZERO {
+        -numerator
+    } else {
+        numerator
+    };
+
+    let positive_divisor = if divisor < T::ZERO { -divisor } else { divisor };
+
+    let (mut quotient, mut remainder) = slow_div_signed(positive_numerator, positive_divisor);
+
+    let numerator_and_divisor_signs_differs = (divisor < T::ZERO) != (numerator < T::ZERO);
+
+    if numerator < T::ZERO {
+        remainder = -remainder;
+    }
+
+    if numerator_and_divisor_signs_differs {
+        quotient = -quotient;
+    }
+
+    (quotient, remainder)
+}
+
+pub(crate) fn absolute_value<T>(value: T) -> T
+where
+    T: Numeric + Neg<Output = T>,
+{
+    if value < T::ZERO {
+        -value
+    } else {
+        value
+    }
+}
+
+pub(crate) enum ShiftType {
+    Logical,
+    Arithmetic,
+}
+
 // move bits from MSB to LSB
-pub(crate) fn shr_assign(lhs: &mut [u64], shift: u32) {
+pub(crate) fn logical_shr_assign(lhs: &mut [u64], shift: u32, shift_type: ShiftType) {
```

Comment:
issue between naming and shift type

Reply:
`shr`is shift right, so naming it still valid to me, a shift right can be either logical or arithemetic

Reply:
shr_assign(... shift_type: ShifType) ?

Reply:
I don't fully get it ? you would prefer renaming `ShiftType`to `RightShiftType`?

Reply:
no what I'm saying is that having a "logical_shr_assign" and giving a type feels weird, I find "shr_assign" with a ShiftType more logical

Reply:
haha my bad, i read too fast, I did not see there was still a `logical` in the name 😅